api/src/main/resources/Java/libraries/okhttp-gson/api.mustache

```diff
@@ -0,0 +1,277 @@
+        return {{localVariablePrefix}}call;
+
+        {{/performBeanValidation}}
+        {{#performBeanValidation}}
+        try {
+            ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
+            ExecutableValidator executableValidator = factory.getValidator().forExecutables();
+
+            Object[] parameterValues = { {{#allParams}}{{paramName}}{{#hasMore}}, {{/hasMore}}{{/allParams}} };
+            Method method = this.getClass().getMethod("{{operationId}}WithHttpInfo"{{#allParams}}, {{#isListContainer}}java.util.List{{/isListContainer}}{{#isMapContainer}}java.util.Map{{/isMapContainer}}{{^isListContainer}}{{^isMapContainer}}{{{dataType}}}{{/isMapContainer}}{{/isListContainer}}.class{{/allParams}});
+            Set<ConstraintViolation<{{classname}}>> violations = executableValidator.validateParameters(this, method,
+                    parameterValues);
+
+            if (violations.size() == 0) {
+                com.squareup.okhttp.Call {{localVariablePrefix}}call = {{operationId}}Call({{#allParams}}{{paramName}}, {{/allParams}}progressListener, progressRequestListener);
+                return {{localVariablePrefix}}call;
+
+            } else {
+                throw new BeanValidationException((Set) violations);
+            }
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
+            throw new ApiException(e.getMessage());
+        } catch (SecurityException e) {
+            e.printStackTrace();
+            throw new ApiException(e.getMessage());
+        }
+
+        {{/performBeanValidation}}
+
+
+
+
+    }
+
+    /**
+     * {{summary}}
+     * {{notes}}{{#allParams}}
+     * @param {{paramName}} {{description}}{{#required}} (required){{/required}}{{^required}} (optional{{#defaultValue}}, default to {{{.}}}{{/defaultValue}}){{/required}}{{/allParams}}{{#returnType}}
+     * @return {{returnType}}{{/returnType}}
+     * @throws ApiException If fail to call the API, e.g. server error or cannot deserialize the response body
+     */
+    public {{#returnType}}{{{returnType}}} {{/returnType}}{{^returnType}}void {{/returnType}}{{operationId}}({{#allParams}}{{{dataType}}} {{paramName}}{{#hasMore}}, {{/hasMore}}{{/allParams}}) throws ApiException {
+        {{! Begin AoU-specific additions }}
+        // Generate an OpenCensus tracing span for this API method call.
+        Scope scope = tracer.spanBuilderWithExplicitParent("{{classname}}.{{operationId}}", tracer.getCurrentSpan())
+            .setSpanKind(Span.Kind.CLIENT).startScopedSpan();
+        Span span = tracer.getCurrentSpan();
+        // Some of these attribute keys use naming from the OpenCensus standard. See
+        // https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/HTTP.md
+        span.putAttribute("package", AttributeValue.stringAttributeValue("{{package}}"));
```

Comment:
Would it make sense to have a utility class to build and populate the span and its attributes and just call that form here? This is a pretty repetitive style. Another approach would be to populate a map here and pass it in.

Even fancier might be a system where you instantiate a span-aware api caller and pass it the code to execute via a lambda. That class could simply be handwritten java, so we don't need to tweak the template as much.

Reply:
I'd originally used a map, but moved back to this style. It's not significantly different, and this ended up being more consistent with the direct span.putAttribute calls in the request-handling and exception-handling cases.